x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/SpatialIntersects.java

```diff
@@ -0,0 +1,249 @@
+    ) {
+        if (spatialDataType == null) {
+            resolveType();
+        }
+
+        EvalOperator.ExpressionEvaluator.Factory leftEvaluator = toEvaluator.apply(left());
+        EvalOperator.ExpressionEvaluator.Factory rightEvaluator = toEvaluator.apply(right());
+        if (isString(right().dataType())) {
+            if (right().foldable()) {
+                // Left is spatial column, right is constant string
+                return new SpatialIntersectsConstantEvaluator.Factory(source(), leftEvaluator, asGeometry(right()));
+            } else {
+                // Left is spatial column, right is variable string
+                return new SpatialIntersectsStringEvaluator.Factory(source(), leftEvaluator, rightEvaluator);
+            }
+        } else {
+            // left and right are spatial columns
+            return new SpatialIntersectsSourceEvaluator.Factory(source(), leftEvaluator, rightEvaluator);
+        }
+    }
+
+    private static BytesRef asGeometry(Expression expression) {
+        Object result = expression.fold();
+        if (result instanceof BytesRef bytesRef) {
+            return SpatialCoordinateTypes.GEO.wktToWkb(bytesRef.utf8ToString());
+        } else {
+            throw new IllegalArgumentException("Invalid constant string: " + result);
+        }
+    }
+
+    @Evaluator(extraName = "Constant")
+    static boolean processConstant(BytesRef leftValue, @Fixed BytesRef rightValue) {
+        return pointIntersectsGeometry(leftValue, rightValue);
+    }
+
+    @Evaluator(extraName = "String")
+    static boolean processString(BytesRef leftValue, BytesRef rightValue) {
+        BytesRef geometry = SpatialCoordinateTypes.GEO.wktToWkb(rightValue.utf8ToString());
+        return pointIntersectsGeometry(leftValue, geometry);
+    }
+
+    @Evaluator(extraName = "Source")
+    static boolean processSource(BytesRef leftValue, BytesRef rightValue) {
+        return geometryIntersectsGeometry(leftValue, rightValue);
+    }
+
+    @Evaluator(extraName = "DocValues")
+    static boolean processDocValues(long leftValue, BytesRef rightValue) {
+        Geometry rightGeom = SpatialCoordinateTypes.GEO.wkbToGeometry(rightValue);
+        return pointIntersectsGeometry(leftValue, rightGeom);
+    }
+
+    private static boolean pointIntersectsGeometry(BytesRef left, BytesRef right) {
+        Geometry leftGeom = SpatialCoordinateTypes.GEO.wkbToGeometry(left);
+        Geometry rightGeom = SpatialCoordinateTypes.GEO.wkbToGeometry(right);
+        if (leftGeom instanceof Point point) {
+            return pointIntersectsGeometry(new GeoPoint(point.getLat(), point.getLon()), rightGeom);
+        } else {
+            throw new IllegalArgumentException(
+                "Can only compare point to other geometries, but was provided left:" + leftGeom.type() + " and right:" + rightGeom.type()
+            );
+        }
+    }
+
+    private static boolean pointIntersectsGeometry(GeoPoint geoPoint, Geometry geometry) {
```

Comment:
My expectations here is that we have a method that transform a Geometry to a component2D and then we can use the method Component2D#contains. I think we have somewhere else already the code that transform a Geometry to a LatLonGeometry or XYGeometry.

Reply:
Now that I've implemented the shape-intersects-shape feature, I have everything necessary to generalize this. However, I think some of these point-specific code paths might still be faster, and worth keeping. We can review this once the code is cleaned up and also once we use your new `LuceneGeometriesUtil`, and done a cleanup because of that.